### ISSUE TITLE
Step by step setup documentation.

### ISSUE THREAD

**Issue by rukai** [google/shaderc-rs#26]
In order to use library with vulkano I need step by step instructions for setting up the build environment in the vulkano readme. It would make sense to also include these instructions in the shaderc-rs readme.

For ubuntu this is easy: `sudo apt-get install git python cmake`
People using other distros can fairly easily adapt this to their own package manager.

I have no access to macOS so I have no idea about this one. I hear brew is popular over there?

For windows I need both msvc and gnu toolchains to work, I'm not sure if they require different tools, however I can only get msvc to work.
I setup the dependencies via msys2. I ran:
`pacman -Syu mingw-w64-x86_64-cmake mingw-w64-x86_64-make mingw-w64-x86_64-gcc mingw64/mingw-w64-x86_64-pkg-config`
and have `C:\msys64\mingw64\bin;C:\msys64\usr\bin` added to my path

At this point I can set the toolchain to msvc and `cargo build` + `cargo test` successfully

However I tried to compile with the gnu toolchain using the terminal window packaged with git and I get the following error with the boilerplate stripped out:
```
$ cargo build
--- stderr
CMake Error at C:/msys64/mingw64/share/cmake-3.12/Modules/CMakeMinGWFindMake.cmake:12 (message):
  sh.exe was found in your PATH, here:

  C:/Program Files/Git/usr/bin/sh.exe

  For MinGW make to work correctly sh.exe must NOT be in your path.

  Run cmake from a shell that does not have sh.exe in your PATH.

  If you want to use a UNIX shell, then use MSYS Makefiles.

Call Stack (most recent call first):
  CMakeLists.txt:8 (project)


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

             at C:\Users\Rukai\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.33\src/lib.rs:773
   7: cmake::run
             at C:\Users\Rukai\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.33\src/lib.rs:754
   8: cmake::Config::build
             at C:\Users\Rukai\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.33\src/lib.rs:606
   9: build_script_build::build_shaderc
             at build/build.rs:21
  10: build_script_build::main
             at build/build.rs:6
```

Alright, so I try with powershell and `cargo build` succeeds but then I run `cargo test` and get so many of these errors that the powershell scrollback is full of them:
```
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_[_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_]+0x21b): undefined reference to `operator delete(void*)'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_[_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_]+0x231): undefined reference to `operator delete(void*)'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_[_ZNSt6vectorIN8spvtools3val10DecorationESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_]+0x2c9): undefined reference to `std::__throw_bad_alloc()'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy[_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy]+0x37): undefined reference to `std::__detail::_Prime_rehash_policy::_M_need_rehash(unsigned long long, unsigned long long, unsigned long long) const'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy[_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy]+0xa9): undefined reference to `operator new(unsigned long long)'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy[_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy]+0x139): undefined reference to `operator delete(void*)'
          C:\Users\Rukai\Desktop\Crates\shaderc-rs\target\debug\build\shaderc-76ef6e510d12bc1e\out\lib\libshaderc_combined.a(validate_instruction.cpp.obj):validate_instruction.cpp:(.text$_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy[_ZNSt10_HashtableIjSt4pairIKjSt3setI17SpvExecutionMode_St4lessIS3_ESaIS3_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toIjESt4hashIjENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb0ELb0ELb1EEEE21_M_insert_unique_nodeEyyPNSA_10_Hash_nodeIS8_Lb0EEEy]+0x1b5): undefined reference to `std::__throw_bad_alloc()'
```

Any idea what I need to do to build with the windows-gnu toolchain?

**Comment by antiagainst** [comment]
Yep, for macOS, I think you typically use Homebrew. 

I'm not quite familiar with MinGW development. Is it a requirement to have both MSVC and GNU toolchain working for Vulkano?

**Comment by rukai** [comment]
Yes, it is a requirement, mostly for selfish reasons. My project requires windows-gnu because libusb only builds there. https://github.com/rukai/PF_Sandbox Of course, the other alternative for me is to get libusb to build on windows-msvc.

I figured you might have a better idea of whats needed to get shaderc building on windows-gnu as I have little experience with C++. However looks like I'll have to dig in and get some experience.

**Comment by nlordell** [comment]
It's a linking issue. Looks like the build script is not linking against stdc++ when targeting windows-gnu. Doing something like this in build.rs gets rid of a lot of errors:

```rust
    match (target_os.as_str(), target_env.as_str()) {
        ("linux", _) | ("windows", "gnu") => println!("cargo:rustc-link-lib=dylib=stdc++"),
        ("macos", _) => println!("cargo:rustc-link-lib=dylib=c++"),
        _ => {}
    }
```

However the link step is still failing complaining about ``undefined reference to `__imp___acrt_iob_func'``. I will keep on investigating

**Comment by nlordell** [comment]
Looks like the ``undefined reference to `__imp___acrt_iob_func'`` errors are a symptom of a larger problem. Take a look at rust-lang/rust#47048 for a longer discussion.

From what I understood it looks like MSYS2 with the latest MinGW-w64 toolchain does not currently work with Rust. I am going to try with mingw-builds toolchain v6.3.0 (same one that Rust uses) and see if I get further.

**Comment by nlordell** [comment]
Using the mingw-builds toolchain v6.3.0 mostly works. `cargo test` successfully runs all tests but fails to compile doc-tests. I think this is linked to rust-lang/rust#41607. From what I gather doc-tests link differently than regular tests and executables. It also seems to affect proc-macro linking.

Additionally linking to:
```rust
    println!("cargo:rustc-link-lib=dylib=pthread");
    println!("cargo:rustc-link-lib=dylib=gcc_eh");
    println!("cargo:rustc-link-lib=dylib=msvcrt");
```
Seems to fix the problem, but I must admit I don't 100% understand what is going on. With this, `cargo test` seems to pass.

**Comment by antiagainst** [comment]
Addressed by #28
